lambdas/backend-api/src/templates/domain/template/template-repository.ts

```diff
@@ -0,0 +1,231 @@
+  UpdateTemplate,
+} from 'nhs-notify-backend-client';
+import {
+  ConditionalCheckFailedException,
+  DynamoDBClient,
+} from '@aws-sdk/client-dynamodb';
+import {
+  DynamoDBDocumentClient,
+  GetCommand,
+  PutCommand,
+  QueryCommand,
+  UpdateCommand,
+  UpdateCommandInput,
+} from '@aws-sdk/lib-dynamodb';
+import { randomUUID as uuidv4 } from 'node:crypto';
+import { ApplicationResult, failure, success } from '@backend-api/utils/result';
+import { Template } from './template';
+
+const client = DynamoDBDocumentClient.from(
+  new DynamoDBClient({ region: 'eu-west-2' }),
+  {
+    marshallOptions: { removeUndefinedValues: true },
+  }
+);
+
+const get = async (
+  templateId: string,
+  owner: string
+): Promise<ApplicationResult<Template>> => {
+  try {
+    const response = await client.send(
+      new GetCommand({
+        TableName: process.env.TEMPLATES_TABLE_NAME,
+        Key: {
+          id: templateId,
+          owner,
+        },
+      })
+    );
+
+    if (!response?.Item) {
+      return failure(ErrorCase.TEMPLATE_NOT_FOUND, 'Template not found');
+    }
+
+    return success(response?.Item as Template);
+  } catch (error) {
+    return failure(ErrorCase.DATABASE_FAILURE, 'Failed to get template', error);
+  }
+};
+
+const create = async (
+  template: CreateTemplate,
+  owner: string
+): Promise<ApplicationResult<Template>> => {
+  const entity: Template = {
+    ...template,
+    id: uuidv4(),
+    owner,
+    version: 1,
+    templateStatus: TemplateStatus.NOT_YET_SUBMITTED,
+    createdAt: new Date().toISOString(),
+    updatedAt: new Date().toISOString(),
+  };
+
+  try {
+    const response = await client.send(
+      new PutCommand({
+        TableName: process.env.TEMPLATES_TABLE_NAME,
+        Item: entity,
+        ReturnConsumedCapacity: 'TOTAL',
+      })
+    );
+
+    const consumedUnits = response?.ConsumedCapacity?.CapacityUnits || 0;
+
+    // Note: This is a bit strange, PutCommand does not return the any data when creating
+    // a new item. So we can only infer the success by checking the CapacityUnits
+    // Or do a GetItem - which is expensive.
+    if (consumedUnits === 0) {
+      throw new Error('Expected DynamoDB CapacityUnits to be greater than 0', {
+        cause: response,
+      });
+    }
+
+    return success(entity);
+  } catch (error) {
+    return failure(
+      ErrorCase.DATABASE_FAILURE,
+      'Failed to create template',
+      error
+    );
+  }
+};
+
+const update = async (
+  templateId: string,
+  template: UpdateTemplate,
+  owner: string
+): Promise<ApplicationResult<Template>> => {
+  const updateExpression = [
+    '#name = :name',
+    '#message = :message',
+    '#updatedAt = :updateAt',
+    '#templateStatus = :templateStatus',
+  ];
+
+  let expressionAttributeNames: Record<string, string> = {
+    '#name': 'name',
+    '#message': 'message',
+    '#templateStatus': 'templateStatus',
+    '#updatedAt': 'updatedAt',
+    '#templateType': 'templateType',
```

Comment:
most of these are not protected words so the use of expressionAttributeNames is not needed